### ISSUE TITLE
Consider if msg-doc / class-doc rules need updating for Vue

### ISSUE THREAD

**Issue by edg2s** [wikimedia/eslint-plugin-mediawiki#11]
I don't know how we'll end up adding constructed classes/message in Vue, but it may well be very different to how we do it currently.

**Comment by catrope** [comment]
The current proposal (implemented in [this patch](https://gerrit.wikimedia.org/r/c/mediawiki/core/+/585797)), for how to use i18n message in Vue provides the following alternatives:
1. For plain text messages: `<p>{{ $i18n( 'msgkey', param1, param2, ... ) }}</p>`
2. For parsed messages: `<p v-i18n-html:msgkey />` or `<p v-i18n-html:msgkey="[param1, param2]" />`
3. For complicated parsed messages: `<p v-i18n-html="{ key: 'msgkey', params: [...] }" />`.

For (1), we could require that the message key passed to `$i18n()` must be a constant, and require dynamic message keys to use a computed property: `<p>{{ complexMessage }}</p>` where `complexMessage: function () { return mw.message( ... ); }`. That way the existing msg-doc rule would kick in and require the possible message keys to be documented.

For (2), we could similarly require that the message key be constant. `<p v-i18n-html:msgkey />` uses the string `'msgkey'` as the message key, but  `<p v-i18n-html:[msgkeyVar] />` uses the value or the variable `msgKeyVar`, and you can even use expressions like `<p v-i18n-html:[foo?bar:baz] />` (as long as the expression doesn't contain spaces or quotes). We would prohibit the use of `<p v-i18n-html:[foo] />` and only allow `<p v-i18n-html:foo />`.

For (3), I think we should discourage (and maybe even prohibit?) writing those object expressions inline, and instead make people use computed properties (i.e. `<p v-i18n-html="complexMessage" />` where `complexMessage: function () { return { key: ..., params: [...] } }`. The downside to this is that it would be hard to detect that the `key:` property is a message key, and the msg-doc rule couldn't enforce documentation. We could address this by renaming `key:` to something more specific like `mwMsgKey:` and have the msg-doc rule look for that, but I think it might be better to change the i18n plugin code to expect a Message object instead. That way, people would have to write `complexMessage: function () { return mw.message( ... ); }` and msg-doc would kick in automatically. We'd just have to prohibit inline usage of `<p v-i18n-html="mw.message( ... )" />`.

In summary, I propose:
- Prohibit passing a dynamic first argument to `$i18n()` in template expressions
- Prohibit passing a dynamic argument to the `v-i18n-html` directive
- Prohibit passing anything other than a simple variable reference to `v-i18n-html` if there is no argument
- Prohibit passing anything other than a simple variable reference or an array to `v-i18n-html` if there is an argument
- Change the i18n plugin to take a Message object instead of `{ key: ..., params: [...] }`

**Comment by catrope** [comment]
For CSS classes, Vue allows you to build the class attribute dynamically using objects or arrays (or strings):
- `<div v-bind:class="{ 'selected': isSelected, 'highlighted: isHighlighted }">`
- `<div v-bind:class="[ 'foo', isCurrent ? 'current' : '' ]">`
- `<div v-bind:class="isRead ? 'read' : 'unread'">`

Again I think we should discourage people from writing complex expressions in templates. Anything more complex than the examples above should probably use a computed property (`<div v-bind:class="classObj">` where `classObj` is a function that returns an object). It probably makes sense to prohibit the use of string concatenation to dynamically build classes in templates (if feasible), but that also means that all cases where string concatenation is used to build class names are in places where we couldn't easily tell that that was the case. So the class-doc rule would be fairly useless, but we could add rules like:

- Warn when `v-bind:class` is an inline object with more than 2 keys, or an array with more than 2 values (we could apply this to `v-bind:style` too, or maybe all `v-bind`s)
- Prohibit string concatenation (somehow? `+` operator where one operand is a string literal?) and template strings in `v-bind:class`

**Comment by catrope** [comment]
I ended up toning down my ambitions a bit. Checking whether something is a "simple variable reference" isn't that easy, because we also want to allow things like `foo.bar`. I ended up just writing a rule that prohibits non-constant first arguments for `$i18n()`, and prohibits passing a dynamic argument to `v-i18n-html`.

I think that's good enough for now. It doesn't always stop you from doing things that are incorrect/broken, but it will always stop you from using dynamic/constructed message keys without documenting them. It does this by preventing the use of any dynamic message keys inside templates, which means you have to use `mw.message()` with a dynamic argument in JS code, which then triggers the `msg-doc` rule.

**Comment by jdforrester** [comment]
Ping; do we want to do anything here?

**Comment by jdforrester** [comment]
Now hopefully fully done in #86 / #90 ; thank you @fflorent!
